### ISSUE TITLE
[SDL_RenderDrawRect] Overlap at corner pixels causes double alpha blending

### ISSUE THREAD

**Issue by pianoman-at-fb** [libsdl-org/SDL#4155]
The rectangle-drawing algorithm seems to have 1 pixel of overlap between the horizontal and vertical lines of the rectangle.  This causes the 4 corner pixels to be alpha-blended twice, which makes them noticeably different from the non-corner pixels.

<img width="624" alt="sdl-bug-drawrect-corners-alpha-minimal" src="https://user-images.githubusercontent.com/58793616/110234303-96a27200-7ede-11eb-9391-abdb69c4e417.png">

See the above screenshot, or compile and run this minimal example:

```c
#include <SDL2/SDL.h>

int main(int argc, char * argv[])
{
    SDL_Init(SDL_INIT_VIDEO);

    SDL_Window * const window = SDL_CreateWindow(
        "DrawRect Corner Alpha Bug",
        SDL_WINDOWPOS_CENTERED,
        SDL_WINDOWPOS_CENTERED,
        512,
        512,
        SDL_WINDOW_RESIZABLE);

    SDL_Renderer * const renderer = SDL_CreateRenderer(
        window,
        -1,
        SDL_RENDERER_ACCELERATED);  // SDL_RENDERER_SOFTWARE also fails

    SDL_RenderSetLogicalSize(renderer, 16, 16);
    SDL_RenderSetIntegerScale(renderer, SDL_TRUE);
    SDL_SetRenderDrawBlendMode(renderer, SDL_BLENDMODE_BLEND);

    SDL_Rect const frame = {.x=4, .y=4, .w=8, .h=8};
    SDL_Event event;

    while(1)
    {
        while(SDL_PollEvent(&event))
        {
            if(SDL_QUIT == event.type)
            {
                SDL_DestroyRenderer(renderer);
                SDL_DestroyWindow(window);
                SDL_Quit();
                return 0;
            }
        }

        SDL_SetRenderDrawColor(renderer, 0, 0, 0, SDL_ALPHA_OPAQUE);
        SDL_RenderClear(renderer);

        SDL_SetRenderDrawColor(renderer, 0x12, 0x66, 0xBB, SDL_ALPHA_OPAQUE/2);
        SDL_RenderDrawRect(renderer, &frame);

        SDL_RenderPresent(renderer);
        SDL_Delay(33);
    }

    return 0;
}
```

On Mac it can be compiled with:

`$ gcc -o bugbugbug sdl2_bug_drawrect_alpha_corners.c -F /Library/Frameworks -framework SDL2`

**Comment by icculus** [comment]
This happens, _only when scaling_, because SDL_RenderDrawRectF breaks this down into 4 line segments to use with SDL_RenderDrawLinesF...but then SDL_RenderDrawLinesF sees that scaling is turned on and does this...

```c
    if (renderer->scale.x != 1.0f || renderer->scale.y != 1.0f) {
        return RenderDrawLinesWithRectsF(renderer, points, count);
    }
```

...and I assume RenderDrawLinesWithRectsF() doesn't consider the endpoints correctly, causing it to alpha blend against stuff in the rectangle we just drew.

Commenting out that code cause the lines to not scale, of course, but if I squint at the one-pixel line, it appears to avoid this, which would make sense, as it would all be part of a single line strip. So that's the likely culprit.

**Comment by icculus** [comment]
(Also, to be clear, this is a separate issue from all the other "line drawing is off by one pixel" bugs and probably exclusively a little incorrect math in RenderDrawLinesWithRectsF)

**Comment by icculus** [comment]
I'm going to kick this to 2.0.20 shortly, since it's separate from the other line rendering issues. The math seems pretty easy to fix this, but I can see a bunch of corner cases I'm going to get wrong.

If we have time between 2.0.18 RC and official release, I'll try to do this one, otherwise, I'll bump to 2.0.20.

**Comment by pianoman-at-fb** [comment]
Thank you for the due diligence @icculus !

**Comment by slouken** [comment]
Fixed for 2.0.20
